File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode* dummy = (struct ListNode*)malloc(sizeof(struct ListNode));
    dummy->next = head;
    struct ListNode* fast = dummy;
    struct ListNode* slow = dummy;

    // Move the fast pointer n+1 steps ahead
    for (int i = 0; i <= n; i++) {
        fast = fast->next;
    }

    // Move both pointers until the fast pointer reaches the end
    while (fast != NULL) {
        fast = fast->next;
        slow = slow->next;
    }

    // Remove the nth node from the end
    struct ListNode* toRemove = slow->next;
    slow->next = slow->next->next;
    free(toRemove);

    struct ListNode* result = dummy->next;
    free(dummy);

    return result;
}

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode fast = head;
        ListNode slow = head;
        if(head == null ) return head;
        if(head.next == null && n == 1) 
        {
            return null;
        }
        
        
        // for(int i = 0;i<n;i++)
        // {
        //     fast = fast.next;
        // }
        int sum = 1;
        while(fast != null){
            if(fast.next == null){
                break;
            }
            if(fast.next.next == null){
                sum = sum + 1;
                break;
            }
          
            sum = sum + 2;
            fast = fast.next.next;
        }
        int rem = sum-n;
        //remove head
        //fast == null
        if(rem == 0)
        {
            return head.next;
        }
 
        // while(fast != null && fast.next != null)
        // {
        //     slow = slow.next;
        //     fast = fast.next;
        // }
        ListNode prev = null;
        // Node front = null;
        for(int i = 0;i<rem;i++){
             
            prev = slow;
            slow = slow.next;
        }
        prev.next = slow.next;
        return head;
    }
}
//BRUTE FORCE USING TWO PASS
// class Solution {
//     public ListNode removeNthFromEnd(ListNode head, int n) {
//         int lengthHead = length( head);
//         ListNode temp = head;
//         int index = lengthHead - n;
//         int count = 0;
//         if(head == null) return head;
//         //remove at head
//         if(index == 0)
//         {
//             return head.next;
//         }
//         while(temp != null)
//         {
//             if(count + 1 == index)
//             {
                 
//                     temp.next = temp.next.next;
//                     break;
//             }
//             count++;
//             temp = temp.next;
//         }
//         return head;

//     }
//     public static int length(ListNode head)
//     {
//         int count = 0;
//         ListNode temp = head;
//         while(temp != null)
//         {
//             count++;
//             temp = temp.next;
//         }
//         return count;
//     }
// }